### ISSUE TITLE
catalog apps - duplicated revisions when deploying an app 

### ISSUE THREAD

**Issue by jiaqiluo** [rancher/rancher#25846]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
- deploy an app from the builtin catalog, in my case it is WordPress
- wait the app the be active, go to rollback the app

**Result:**
- there are 2 duplicated revisions on the list
<img width="977" alt="Screen Shot 2020-03-06 at 4 23 06 PM" src="https://user-images.githubusercontent.com/6218999/76130718-0d094d00-5fc9-11ea-9919-52ff0f64ca5b.png">

**Other details that may be helpful:**
- this does not happen to the multi-cluster apps 

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI):
master-head 63306835a
- Installation option (single install/HA):
single install

**Comment by jiaqiluo** [comment]
the same issue happens for the WordPress app from the builtin `library` catalog which uses helm 2.

Steps:
- go to a project
- deploy the `WordPress` app from the builtin `library` catalog 
- check revisions after the app is active

**Comment by jiaqiluo** [comment]
See the same issue when using a helm 3 catalog

steps:
- go to a project
- add this catalog with helm 3 `https://argoproj.github.io/argo-helm`
![image](https://user-images.githubusercontent.com/6218999/76245971-1decd500-61fa-11ea-904c-8b2278786978.png)

- deploy the app `argo` from the catalog 
- check revisions after the app is active
![image](https://user-images.githubusercontent.com/6218999/76246027-352bc280-61fa-11ea-9611-8649fa472feb.png)

results:
- two revisions show in the list and they are identical 
![image](https://user-images.githubusercontent.com/6218999/76246050-3c52d080-61fa-11ea-8f53-dc17ed4c7a4c.png)

**Comment by brendarearden** [comment]
Running from this commit  `6e723b723`  prior to helm changes, I am seeing this same behavior for helm 2 to, so it seems like something else introduced after v2.3.5 that is causing these issues unrelated to helm 3 changes.

**Comment by brendarearden** [comment]
**TLDR**
Changing to an api call instead of using cache does not solve root problem of how we are checking for app obj changes. Need to resolve how changes are being detected

**Initial fix attempt**
So I initially attempted to fix this by switching from the `AppsLister` to the `AppsGetter` to avoid using the cache and attempt to get the obj from the api directly. The downside of this would be an exponential number of api calls for customers with a large amount of apps deployed and would not address the root cause, it would just decrease the window for a race condition. 

**Scenarios and outcomes discovered during initial fix attempt**
1. Use AppsLister: results in duplicate apps on app install
2. Use AppsGetter only when AppRevisionName == "" : results in duplicate app revisions on upgrade
3. Use AppsGetter: results in no duplicate apps but causes performance issue exponentially 

**Root Cause**
For each app change, an AppRevision is created that a user can roll back to if needed. The name of this AppRevision is being stored on the app obj. As the app obj is running through an update, it is randomly resulting in a race condition because the AppRevisionName is not updating on the app obj. This is because the app revision is created right before update completes and starts over again so the cache isn't updated in time. This is resulting in an additional revision being created as it is not able to verify current obj against the app revision

**Potential Solutions**
1) 
We are using the AppRevision as the single source of truth and comparing the current app object with the most recent app revision to check for changes. An app revision should be a snapshot in time and potentially shouldn't be our source of truth. The current idea for this solution is that we should be checking the obj's status with its spec to see if changes are being made. The status should be the existing obj status and the spec should be the user requested changes. With this solution we will need to ensure that any changes can also be caught on existing deployed / old apps
2) 
see this comment https://github.com/rancher/rancher/pull/27179#issuecomment-632812061 about MutationCache as an alternative to ensure we have the most up to date changes stored locally even if the cache hasn't update yet

Feel free to make any suggestions or point out any incorrect assumptions ^^ :)
